src/org/thoughtcrime/securesms/ConversationFragment.java

```diff
@@ -520,12 +533,13 @@ public long stageOutgoingMessage(OutgoingMediaMessage message) {
     public long stageOutgoingMessage(OutgoingTextMessage message) {
         MessageRecord messageRecord = DatabaseFactory.getSmsDatabase(getContext()).readerFor(message, threadId).getCurrent();
 
-        if (getListAdapter() != null) {
-            getListAdapter().setHeaderView(null);
-            setLastSeen(0);
-            getListAdapter().addFastRecord(messageRecord);
+        if (! (messageRecord.getBody().getBody().length() >= 19 && messageRecord.getBody().getBody().substring(0, 19).equals("{\"type\": \"reaction\""))) {
```

Comment:
I don't like that this check is being repeated again here (like in `ConversationActivity`). I think it would be a interesting idea to have an external parsing class which could maybe event return the exact data type of the message.

```java
String type = Stereotyper.getTypeFromBody(messageRecord.getBody());
if (type == null) {
    // do normal stuff
} if (type == MessageTypes.REACTION)  {
    // do reaction stuff
}
```

(would also be something that is easy to unit test)